src/storage-controller/src/lib.rs

```diff
@@ -1197,41 +1130,89 @@ where
-        self.drop_sources_unvalidated(storage_metadata, identifiers);
-        Ok(())
+        self.drop_sources_unvalidated(storage_metadata, identifiers)
+            .await
     }
 
-    fn drop_sources_unvalidated(
+    async fn drop_sources_unvalidated(
         &mut self,
         storage_metadata: &StorageMetadata,
-        identifiers: Vec<GlobalId>,
-    ) {
-        for id in &identifiers {
-            let metadata = storage_metadata.get_collection_shard::<T>(*id);
+        ids: Vec<GlobalId>,
+    ) -> Result<(), StorageError<Self::Timestamp>> {
+        let mut to_execute = BTreeSet::new();
+        let mut to_drop = BTreeSet::new();
+        for id in ids {
+            let metadata = storage_metadata.get_collection_shard::<T>(id);
             mz_ore::soft_assert_or_log!(
                 matches!(metadata, Err(StorageError::IdentifierMissing(_))),
                 "dropping {id}, but drop was not synchronized with storage \
                 controller via `synchronize_collections`"
             );
+
+            let dropped_data_source = match self.collection(id) {
+                Ok(col) => col.description.data_source.clone(),
+                Err(_) => continue,
+            };
+
+            to_drop.insert(id);
+
+            // If we are dropping source exports, we need to modify the
+            // ingestion that it runs on.
+            if let DataSource::IngestionExport { ingestion_id, .. } = dropped_data_source {
+                // Adjust the source to contain this export.
+                let ingestion_collection = match self.collections.get_mut(&ingestion_id) {
+                    Some(ingestion_collection) => ingestion_collection,
+                    // Primary ingestion already dropped.
+                    None => {
+                        tracing::error!(
+                            "primary source {ingestion_id} seemingly dropped before subsource {id}",
+                        );
+                        continue;
+                    }
+                };
+
+                match &mut ingestion_collection.description {
+                    CollectionDescription {
+                        data_source: DataSource::Ingestion(ingestion_desc),
+                        ..
+                    } => {
+                        let removed = ingestion_desc.source_exports.remove(&id);
+                        mz_ore::soft_assert_or_log!(
+                            removed.is_some(),
+                            "dropped subsource {id} already removed from source exports"
+                        );
+                    }
+                    _ => unreachable!(
+                        "SourceExport must only refer to primary sources that already exist"
+                    ),
+                };
+            }
+        }
+
+        // Do not bother re-executing collections we know we plan to drop.
+        to_execute.retain(|id| !to_drop.contains(id));
```

Comment:
we never push anything into `to_execute`. What is going on here?

Reply:
😮‍💨 this got lost in a rebase; if we remove a source export, we should re-execute its ingestion.